Allow rpc.remote to create RRef on self

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29543 [WIP] Add end-to-end test for RNN Module
* #29781 Always include autograd context id in rpc/remote requests
* **#29634 Allow rpc.remote to create RRef on self**

This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

mrshenli committed:
Allow rpc.remote to create RRef on self

This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

[ghstack-poisoned]

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

aazzolini commented on test/rpc_test.py:
```diff
@@ -237,6 +237,12 @@ def test_self_add(self):
         self.assertEqual(fut.wait(), torch.ones(2, 2) + 1)
         self.assertEqual(ret, torch.ones(2, 2) + 1)
 
+    @dist_init
```
We also need to test that we can pass RRef's as arguments to local sync_rpc, async_rpc and remote calls.

mrshenli replied:
yes, will add more tests

aazzolini commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -171,29 +178,46 @@ PyRRef pyRemotePythonUdf(
+    fm->addCallback(finishAcceptUserRRef);
+    return PyRRef(userRRef);
+  } else {
+    auto ownerRRef = ctx.createOwnerRRef<py::object>(true /* retainInContext*/);
```
retainInContext is hard to understand. would it make sense to make this a separate function call?

mrshenli replied:
Sure, I am working on refactoring this PR, will send out an update today.

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

aazzolini commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -96,7 +96,11 @@ Message RequestCallbackImpl::processRpc(
-      ctx.addForkOfOwner(rrefId, forkId);
+
+      if (rrefId != forkId) {
+        // This is not owner calling to self
```
Can you add a more specific comment here on where we create a fork where rrefId == forkId (namely , when calling "remote" to self).

aazzolini commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -142,7 +142,7 @@ template std::shared_ptr<OwnerRRef<py::object>> RRefContext::
 template <typename T>
 std::shared_ptr<OwnerRRef<T>> RRefContext::createOwnerRRef() {
-  // Don't add this OnwerRRef to the owners_ map yet, otherwise
+  // Don't add this OnwerRRef to the owners_ map yet by default, otherwise
```
is this change in the comment still pertinent in the latest version?

mrshenli replied:
This change should be removed, let me update. Thanks!

aazzolini reviewed (approved):
Looks good. In the future we should probably bypass serialization and deserialization and simply short-circuit the local call, but that will require access to the thread pool.

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]

mrshenli committed:
Update on "Allow rpc.remote to create RRef on self"


This implementation supports rpc.remote to self by doing the
following steps:

1. create an owner RRef
2. add the owner RRef to owners_ in RRefContext, and keep it alive
   by using RRefId as the ForkId.
3. Go through serde and insert the message to the caller's thread-pool
4. When the response message gets processed, remove the itself from
   RRef fork map.

Differential Revision: [D18445812](https://our.internmc.facebook.com/intern/diff/D18445812)

[ghstack-poisoned]
